api/agent/lb_agent.go

```diff
@@ -177,10 +208,40 @@ func (a *lbAgent) Submit(callI Call) error {
 	statsDequeue(ctx)
 	statsStartRun(ctx)
 
-	err = a.placer.PlaceCall(a.rp, ctx, call)
+	if call.Type == models.TypeDetached {
+		return a.placeDetachCall(ctx, call)
+	}
+	return a.placeCall(ctx, call)
+}
+
+func (a *lbAgent) placeDetachCall(ctx context.Context, call *call) error {
+	errPlace := make(chan error, 1)
+	rw := call.w.(*DetachedResponseWriter)
+	go a.spawnPlaceCall(ctx, call, errPlace)
+	select {
+	case err := <-errPlace:
+		return err
+	case <-rw.acked:
+		return nil
+	}
+}
+
+func (a *lbAgent) placeCall(ctx context.Context, call *call) error {
+	err := a.placer.PlaceCall(a.rp, ctx, call)
 	return a.handleCallEnd(ctx, call, err, true)
 }
 
+func (a *lbAgent) spawnPlaceCall(ctx context.Context, call *call, errCh chan error) {
+	var cancel func()
+	ctx = common.BackgroundContext(ctx)
+	// 30 secs PlacerTimeout for Detached + call.Timeout (inside container) + 360 sec headroom
+	// to make sure we do not wait indefinitely, but also wait enough time for worst case.
+	ctx, cancel = context.WithTimeout(ctx, time.Duration(360+call.Timeout)*time.Second)
```

Comment:
360 is kinda a magic number here. this doesn't respect the cancellation of the calling request context, which may cancel or have a timeout associated with it itself (remove `common.BackgroundContext`). this seems like the right place to impose `DetachedPlacerTimeout`, really, and let everything underneath consume that. this goes back to an earlier comment about plumbing, if we make the Submit contract have an additional surface area for returning early on 202, stuff like this kind of falls out (PlaceCall would return after getting the call detached message, but before the task is finished, rather than us having to guess how long it might take). my real question would be for PlaceCall for detached can we make it just return earlier for detached rather than having 2 different calling paths? this would make it 'just fall out' and return from placeCall to the http handler after the call was 'placed' but not executed in 'detached' mode?

Reply:
I refactored the code here, please havea look and see if it makes sense. 
In my understanding at this point  we have to create a new context and we can't reuse the original context coming with the request as the original request will be closed as soon as we return the 202 and that will cause the abort of the running function. Is my understanding correct?

@skinowski as regards the new timeout of the new context I think it should be call.Timeout + DetachPlacerTimeout, again I don't see why we should add extra room here and we don't do the same for the regular sync call

Reply:
As @reclaro says, we should not respect caller ctx timeout here. This is "background" detached mode in LB. It means, the request is being run as if client is still present, hence common.BackgroundContext(). We no longer care about client ctx basically. 

Here the rationale is this:

* 30 sec placer timeout means, LB might wait up to 30 secs waiting to push the call to a runner
* call.Timeout means time inside a container
* 360 headroom means anything else (docker-pull, runner scheduler delay, gRPC/network retransmit, etc. etc.)

New changes are coming in runner soon to bound the docker-pull and other latencies. So, this 360 needs to be in coordination of that. We should move the constant to a config/env var.

The code however is not in sync with the comments. Math there seems like "360 + call.Timeout" which omits placer timeout.

Reply:
placer timeout from the update makes sense, thanks. 

i'm not sure that's the case wrt ctx cancel with lb_agent/pure_runner, they have a protocol to handle moving the call state forward that is more 'stateless' in nature. in lb_agent using the request context won't cancel a task running in the pure runner, as far as I know, I only see one handle of cancellation in lb_agent and it doesn't shut down the task (my understanding is, it would have to handle it explicitly to do this, and I don't see it's doing that). this just means the placer timeout is imposing a server side configuring client timeout for detached tasks, basically. while we do need to bound it, it's interesting to think about just allowing the http request to specify this timeout (it's not a hard req or blocking here, what's here with placer timeout looks ok wrt timings)

Reply:
lb_agent doesn't appear have timeouts around each state machine for each call, so i'm not sure the assertions about ignoring request context are true. in any case, the 360 number really what we want is 30 here? and lb_agent shouldn't be coordinating timeouts with the runner at all because the protocol doesn't time out or cancel any individual tasks? not that the protocol is well defined in that respect